helix-term/src/commands/dap.rs

```diff
@@ -0,0 +1,871 @@
+            for (i, x) in params.iter().enumerate() {
+                let mut param = x.to_string();
+                if let Some(DebugConfigCompletion::Advanced(cfg)) = template.completion.get(i) {
+                    if matches!(cfg.completion.as_deref(), Some("filename" | "directory")) {
+                        param = std::fs::canonicalize(x)
+                            .ok()
+                            .and_then(|pb| pb.into_os_string().into_string().ok())
+                            .unwrap_or_else(|| x.to_string());
+                    }
+                }
+                // For param #0 replace {0} in args
+                let pattern = format!("{{{}}}", i);
+                value = match value {
+                    // TODO: just use toml::Value -> json::Value
+                    DebugArgumentValue::String(v) => {
+                        DebugArgumentValue::String(v.replace(&pattern, &param))
+                    }
+                    DebugArgumentValue::Array(arr) => DebugArgumentValue::Array(
+                        arr.iter().map(|v| v.replace(&pattern, &param)).collect(),
+                    ),
+                    DebugArgumentValue::Boolean(_) => value,
+                };
+            }
+
+            match value {
+                DebugArgumentValue::String(string) => {
+                    if let Ok(integer) = string.parse::<usize>() {
+                        args.insert(k, to_value(integer).unwrap());
+                    } else {
+                        args.insert(k, to_value(string).unwrap());
+                    }
+                }
+                DebugArgumentValue::Array(arr) => {
+                    args.insert(k, to_value(arr).unwrap());
+                }
+                DebugArgumentValue::Boolean(bool) => {
+                    args.insert(k, to_value(bool).unwrap());
+                }
+            }
+        }
+    }
+
+    let args = to_value(args).unwrap();
+
+    let callback = |_editor: &mut Editor, _compositor: &mut Compositor, _response: Value| {
+        // if let Err(e) = result {
+        //     editor.set_error(format!("Failed {} target: {}", template.request, e));
+        // }
+    };
+
+    match &template.request[..] {
+        "launch" => {
+            let call = debugger.launch(args);
+            dap_callback(cx.jobs, call, callback);
+        }
+        "attach" => {
+            let call = debugger.attach(args);
+            dap_callback(cx.jobs, call, callback);
+        }
+        request => bail!("Unsupported request '{}'", request),
+    };
+
+    // TODO: either await "initialized" or buffer commands until event is received
+    cx.editor.debugger = Some(debugger);
+    let stream = UnboundedReceiverStream::new(events);
+    cx.editor.debugger_events.push(stream);
+    Ok(())
+}
+
+pub fn dap_launch(cx: &mut Context) {
+    if cx.editor.debugger.is_some() {
+        cx.editor
+            .set_error("Debugger is already running".to_string());
+        return;
+    }
```

Comment:
`editor.set_error(...); return;` is becoming a common enough pattern across commands so much that I think it's better to have commands return a Result, similar to typable commands  and handle setting the error in a single place. Better off as a separate PR though.

Reply:
I've been thinking about this but I'm not sure. DAP/LSP commands need error handling, but most built in commands don't. I guess it would be nice for consistency?